File: lib/route-finder.ts
import { Graph, RouteResult, RouteStep, StationsMap } from '@/types/metro';

export function findRoutes(
  graph: Graph,
  stations: StationsMap,
  from: string,
  to: string,
  maxRoutes: number = 5
): RouteResult[] {
  if (from === to) return [];

  const allRoutes: RouteResult[] = [];

  // Use DFS with path tracking to find multiple routes
  const findAllPaths = (
    current: string,
    target: string,
    visited: Set<string>,
    path: string[],
    lines: string[], // Track lines used in path
    currentLine: string,
    transfers: number
  ) => {
    if (current === target) {
      // Build route from path
      const steps: RouteStep[] = [];
      let prevLine = '';

      for (let i = 0; i < path.length; i++) {
        const stationId = path[i];
        const station = stations[stationId];
        if (!station) continue;

        const edge = i > 0 ? graph[path[i - 1]]?.find(e => e.to === stationId) : null;
        const line = edge?.line || station.lines?.[0] || '';
        const isTransfer = line !== prevLine && prevLine !== '';

        let transferTo: string | undefined;
        if (isTransfer) {
          transferTo = line;
        }

        steps.push({
          stationId,
          station,
          line,
          isTransfer,
          transferTo,
        });

        prevLine = line;
      }

      const uniqueLines = Array.from(new Set(steps.map(s => s.line).filter(Boolean)));

      allRoutes.push({
        steps,
        totalStations: path.length,
        totalTransfers: uniqueLines.length - 1,
        lines: uniqueLines,
      });

      return;
    }

    // Stop if path is too long or too many transfers
    if (path.length > 30 || transfers > 4) return;

    const edges = graph[current] || [];

    // Sort edges to prioritize continuing on same line
    const sortedEdges = [...edges].sort((a, b) => {
      const aIsSameLine = a.line === currentLine ? 0 : 1;
      const bIsSameLine = b.line === currentLine ? 0 : 1;
      return aIsSameLine - bIsSameLine;
    });

    for (const edge of sortedEdges) {
      if (!visited.has(edge.to) && stations[edge.to]) {
        const newTransfers = currentLine && edge.line !== currentLine ? transfers + 1 : transfers;
        const newVisited = new Set(visited);
        newVisited.add(edge.to);

        const newLines = edge.line !== currentLine ? [...lines, edge.line] : lines;

        findAllPaths(
          edge.to,
          target,
          newVisited,
          [...path, edge.to],
          newLines,
          edge.line,
          newTransfers
        );
      }
    }
  };

  const firstStation = stations[from];
  const initialLine = firstStation?.lines?.[0] || '';
  const initialVisited = new Set<string>([from]);

  findAllPaths(from, to, initialVisited, [from], [initialLine], initialLine, 0);

  // Remove duplicate routes (same stations in same order)
  const uniqueRoutes = allRoutes.filter((route, index, self) => {
    const pathKey = route.steps.map(s => s.stationId).join('-');
    return index === self.findIndex(r => r.steps.map(s => s.stationId).join('-') === pathKey);
  });

  // Sort by transfers first (PRIORITY), then by total stations
  uniqueRoutes.sort((a, b) => {
    // First compare by number of transfers
    if (a.totalTransfers !== b.totalTransfers) {
      return a.totalTransfers - b.totalTransfers;
    }
    // If transfers are equal, compare by total stations
    return a.totalStations - b.totalStations;
  });

  // Return top routes
  return uniqueRoutes.slice(0, maxRoutes);
}

File: apps/android/app/src/main/java/ir/tehgo/routing/RouteFinder.kt
package ir.tehgo.routing

import ir.tehgo.data.MetroGraph
import ir.tehgo.data.Station

/**
 * Ported from web-version/lib/route-finder.ts: DFS multi-route search over the metro graph,
 * capped at 30 stations / 4 transfers per path, sorted by transfers then station count.
 */
object RouteFinder {

    fun findRoutes(
        graph: MetroGraph,
        stationsById: Map<String, Station>,
        from: String,
        to: String,
        maxRoutes: Int = 5,
    ): List<RouteResult> {
        if (from == to) return emptyList()

        val allRoutes = mutableListOf<RouteResult>()

        fun buildRoute(path: List<String>): RouteResult {
            val steps = mutableListOf<RouteStep>()
            var prevLine = ""

            for (i in path.indices) {
                val stationId = path[i]
                if (stationsById[stationId] == null) continue

                val edge = if (i > 0) graph[path[i - 1]]?.find { it.to == stationId } else null
                val line = edge?.line ?: stationsById[stationId]?.lineIds?.firstOrNull().orEmpty()
                val isTransfer = line != prevLine && prevLine.isNotEmpty()

                if (isTransfer && steps.isNotEmpty()) {
                    steps[steps.size - 1] = steps.last().copy(transferTo = line)
                }

                steps += RouteStep(stationId = stationId, line = line, isTransfer = isTransfer)
                prevLine = line
            }

            val uniqueLines = steps.map { it.line }.filter { it.isNotEmpty() }.distinct()
            return RouteResult(
                steps = steps,
                totalStations = path.size,
                totalTransfers = uniqueLines.size - 1,
                lines = uniqueLines,
            )
        }

        val visited = HashSet<String>().apply { add(from) }
        val path = ArrayList<String>().apply { add(from) }

        fun dfs(current: String, currentLine: String, transfers: Int) {
            // Bounded early exit: once we've collected a generous oversample of candidates
            // (maxRoutes * 6), stop exploring further. Routes are sorted by
            // (totalTransfers, totalStations) and truncated to maxRoutes after the search
            // completes, so oversampling well past maxRoutes before stopping keeps the final
            // top-maxRoutes result essentially certain to match an exhaustive search - this
            // only bounds worst-case work on pathological/future-larger graphs, it isn't an
            // exact top-K algorithm.
            if (allRoutes.size >= maxRoutes * 6) return
            if (current == to) {
                allRoutes += buildRoute(path)
                return
            }
            if (path.size > 30 || transfers > 4) return

            val edges = (graph[current] ?: emptyList()).sortedBy { if (it.line == currentLine) 0 else 1 }
            for (edge in edges) {
                if (edge.to !in visited && stationsById.containsKey(edge.to)) {
                    val newTransfers = if (currentLine.isNotEmpty() && edge.line != currentLine) transfers + 1 else transfers
                    visited += edge.to
                    path += edge.to
                    dfs(edge.to, edge.line, newTransfers)
                    path.removeAt(path.size - 1)
                    visited -= edge.to
                }
            }
        }

        val initialLine = stationsById[from]?.lineIds?.firstOrNull().orEmpty()
        dfs(from, initialLine, 0)

        val uniqueRoutes = allRoutes.distinctBy { route -> route.steps.joinToString("-") { it.stationId } }
        return uniqueRoutes
            .sortedWith(compareBy({ it.totalTransfers }, { it.totalStations }))
            .take(maxRoutes)
    }
}
